File: core/ReverseNodesInKGroup/ReverseNodesInKGroup.java
package com.dbc;

import java.util.List;

public class ReverseNodesInKGroup {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    private void reverseList(ListNode preHead, ListNode endNode) {
        ListNode head = preHead.next;
        preHead.next = endNode;
        ListNode preNode = endNode.next;
        endNode.next = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = preNode;
            preNode = head;
            head = temp;
        }
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode preHead = new ListNode();
        preHead.next = head;
        ListNode left = preHead, right = head;
        int count = 1;
        while (right != null) {
            while (right != null && count < k) {
                right = right.next;
                count++;
            }
            if (right != null && count == k) {
                ListNode temp = left.next;
                reverseList(left, right);
                count = 1;
                left = temp;
                right = left.next;
            }
        }
        return preHead.next;
    }
}

File: core/ReverseNodesInKGroup/ReverseNodesInKGroup.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverse_list(pre_head, end_node):
    head = pre_head.next
    pre_head.next = end_node
    pre_node = end_node.next
    end_node.next = None
    while head is not None:
        temp_node = head.next
        head.next = pre_node
        pre_node = head
        head = temp_node


def reverseKGroup(head: Optional[ListNode], k: int) -> Optional[ListNode]:
    count = 1
    prev_head = ListNode()
    prev_head.next = head
    left, right = prev_head, head
    while right is not None:
        while right is not None and count < k:
            right = right.next
            count += 1

        if right is not None and count == k:
            temp = left.next
            reverse_list(left, right)
            count = 1
            left = temp
            right = left.next

    return prev_head.next


